dotCMS/src/main/java/com/dotcms/filters/interceptor/saml/SamlWebInterceptor.java

```diff
@@ -0,0 +1,682 @@
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpSession;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.dotcms.saml.DotSamlConstants.SAML_USER_ID;
+
+/**
+ * This interceptor encapsulates the logic for Saml
+ * Basically if there is any configuration set on apps and c
+ * @author jsanca
+ */
+public class SamlWebInterceptor implements WebInterceptor {
+
+    public static final String APPS_SAML_CONFIG_KEY   = "app-saml-config";
+    public static final String REFERRER_PARAMETER_KEY = "referrer";
+    public static final String ORIGINAL_REQUEST       = "original_request";
+    public static final String BY_PASS_KEY            = "native";
+    public static final String BY_PASS_VALUE          = "true";
+
+    protected final Encryptor       encryptor;
+    protected final LoginServiceAPI loginService;
+    protected final UserAPI         userAPI;
+    protected final UserWebAPI      userWebAPI;
+    protected final HostWebAPI      hostWebAPI;
+    protected final AppsAPI         appsAPI;
+    protected final IdentityProviderConfigurationFactory identityProviderConfigurationFactory;
+    protected final SamlConfigurationService             samlConfigurationService;
+    protected final SamlAuthenticationService            samlAuthenticationService;
+
+    public SamlWebInterceptor() {
+        this(APILocator.getLoginServiceAPI());
+    }
+
+    public SamlWebInterceptor(final LoginServiceAPI loginServiceAPI) {
+        this.loginServiceAPI = loginServiceAPI;
+    }
+
+    @Override
+    public Result intercept(final HttpServletRequest request,
+                            final HttpServletResponse response) throws IOException {
+
+        HttpSession session = request.getSession(false);
+
+        if (this.isByPass(request, session)) {
+
+            Logger.info(this, ()->"Using SAML by pass");
+            return Result.NEXT;
+        }
+
+        Result result      = Result.NEXT;
+
+        try {
+
+            final boolean useSaml      = this.isAnySamlConfigurated();
+            String redirectAfterLogin  = null;
+
+            if (useSaml && null != session) {
+
+                final Host host = hostWebAPI.getCurrentHostNoThrow(request);
+                final IdentityProviderConfiguration identityProviderConfiguration =
+                        this.identityProviderConfigurationFactory.findIdentityProviderConfigurationById(
+                                host.getIdentifier());
+
+                // If idpConfig is null, means this site does not need SAML processing
+                if (null != identityProviderConfiguration && identityProviderConfiguration.isEnabled()) {
+
+                    // SAML is configurated
+                    final boolean isLogoutNeed = this.samlConfigurationService.getConfigAsBoolean(
```

Comment:
Codacy found an issue: [Avoid declaring a variable if it is unreferenced before a possible exit point.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5707914)